core/capabilities/syncer.go

```diff
@@ -53,172 +67,647 @@ var defaultStreamConfig = p2ptypes.StreamConfig{
+			if err != nil {
+				return fmt.Errorf("failed to add trigger shim: %w", err)
 			}
-			workflowDONs := map[string]capabilities.DON{
-				s.networkSetup.WorkflowsDonInfo.ID: s.networkSetup.WorkflowsDonInfo,
+		case capabilities.CapabilityTypeAction:
+			s.lggr.Warn("no remote client configured for capability type action, skipping configuration")
+		case capabilities.CapabilityTypeConsensus:
+			s.lggr.Warn("no remote client configured for capability type consensus, skipping configuration")
+		case capabilities.CapabilityTypeTarget:
+			newTargetFn := func(info capabilities.CapabilityInfo) (capabilityService, error) {
+				client := target.NewClient(
+					info,
+					*toDONInfo(remoteDON),
+					s.dispatcher,
+					defaultTargetRequestTimeout,
+					s.lggr,
+				)
+				return client, nil
 			}
-			triggerCap := remote.NewTriggerPublisher(config, underlying, triggerInfo, s.networkSetup.TriggerCapabilityDonInfo, workflowDONs, s.dispatcher, s.lggr)
-			err = s.dispatcher.SetReceiver(capId, s.networkSetup.TriggerCapabilityDonInfo.ID, triggerCap)
+
+			err := s.addToRegistryAndSetDispatcher(ctx, capability, remoteDON, newTargetFn)
 			if err != nil {
-				s.lggr.Errorw("capability DON failed to set receiver", "capabilityId", capId, "donId", s.networkSetup.TriggerCapabilityDonInfo.ID, "error", err)
-				return
+				return fmt.Errorf("failed to add target shim: %w", err)
 			}
-			s.subServices = append(s.subServices, triggerCap)
-			break
+		default:
+			s.lggr.Warnf("unknown capability type, skipping configuration: %+v", capability)
 		}
 	}
-	if s.networkSetup.IsTargetDon(myId) {
-		s.lggr.Info("member of a target DON - starting remote shims")
-		underlying, err2 := s.registry.GetTarget(ctx, targetCapId)
-		if err2 != nil {
-			s.lggr.Errorw("target not found yet", "capabilityId", targetCapId, "error", err2)
-			return
+	return nil
+}
+
+type capabilityService interface {
+	capabilities.BaseCapability
+	remotetypes.Receiver
+	services.Service
+}
+
+func (s *registrySyncer) addToRegistryAndSetDispatcher(ctx context.Context, capabilityInfo kcr.CapabilityRegistryCapability, don kcr.CapabilityRegistryDONInfo, newCapFn func(info capabilities.CapabilityInfo) (capabilityService, error)) error {
+	fullCapID := fmt.Sprintf("%s@%s", capabilityInfo.LabelledName, capabilityInfo.Version)
+	info, err := capabilities.NewRemoteCapabilityInfo(
+		fullCapID,
+		toCapabilityType(capabilityInfo.CapabilityType),
+		fmt.Sprintf("Remote Capability for %s", fullCapID),
+		toDONInfo(don),
+	)
+	if err != nil {
+		return err
+	}
+	capability, err := newCapFn(info)
+	if err != nil {
+		return err
+	}
+
+	err = s.registry.Add(ctx, capability)
+	if err != nil {
+		// If the capability already exists, then it's either local
+		// or we've handled this in a previous syncer iteration,
+		// let's skip and move on to other capabilities.
+		if errors.Is(err, ErrCapabilityAlreadyExists) {
+			return nil
```

Comment:
If the capability is local (or was not added through the syncer), is it guaranteed to have a receiver and be started as a subservice (steps below)? Is it possible that this may return too early?

Reply:
I sent you a response on slack: the tl;dr is that this will be covered by enableExternalAccess